Fix issue 11225: Improve CustomElementRegistry.define() page (#29317)
mdn/content: files/en-us/web/api/customelementregistry/define/index.md

### Exceptions

- `NotSupportedError` {{domxref("DOMException")}}
  - : Thrown if:
    - The {{domxref("CustomElementRegistry")}} already contains an entry with the same name or the same constructor (or is otherwise already defined).
    - The <code>extends</code> option is specified and it is a [valid custom element name](#valid_custom_element_names)
    - The <code>extends</code> option is specified but the element it is trying to extend is an unknown element.
- `SyntaxError` {{domxref("DOMException")}}
  - : Thrown if the provided name is not a [valid custom element name](#valid_custom_element_names).
- {{jsxref("TypeError")}}
  - : Thrown if the referenced constructor is not a constructor.

## Description

The `define()` method adds a definition for a custom element to the custom element registry, mapping its name to the constructor which will be used to create it.

There are two types of custom element you can create:

> - _Autonomous custom element_ are standalone elements, that don't inherit from built-in HTML elements.
> - _Customized built-in element_ are elements that inherit from, and extend, built-in HTML elements.
> 

Reviewer: ```suggestion
- _Autonomous custom elements_ are standalone elements, that don't inherit from built-in HTML elements.
- _Customized built-in elements_ are elements that inherit from, and extend, built-in HTML elements.

```

To define an autonomous custom element, you should omit the `options` parameter.

> To define a customized built-in element, you must pass the `options` parameter with its `extends` property set to the name of the built-in element that you are extending, and this must correspond to the interface that your custom element class definition inherits. For example, to customize the {{htmlelement("p")}} element, you must pass `{extends: "p"}` to `define()`, and the class definition for your element must inherit from {{domxref("HTMLParagraphElement")}}.

Reviewer: ```suggestion
To define a customized built-in element, you must pass the `options` parameter with its `extends` property set to the name of the built-in element that you are extending, and this must correspond to the interface that your custom element class definition inherits from. For example, to customize the {{htmlelement("p")}} element, you must pass `{extends: "p"}` to `define()`, and the class definition for your element must inherit from {{domxref("HTMLParagraphElement")}}.
```

### Valid custom element names

Custom element names must:

> - start with an ASCII lower-case letter (a-z)
> - contain a hypen
> - not contain any ASCII upper-case letters

Reviewer: ```suggestion
- start with an ASCII lowercase letter (a-z)
- contain a hypen
- not contain any ASCII uppercase letters
```
